pkg/yttlibrary/assert.go

```diff
@@ -25,6 +28,69 @@ var (
 
 type assertModule struct{}
 
+func (b assertModule) Equals(thread *starlark.Thread, f *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	if args.Len() != 2 {
+		return starlark.None, fmt.Errorf("expected two arguments")
+	}
+
+	expected := args.Index(0)
+	actual := args.Index(1)
+	if _, notOk := expected.(starlark.Callable); notOk {
+		return starlark.None, fmt.Errorf("expected argument not to be a function, but was %T", expected)
+	}
+	if _, notOk := actual.(starlark.Callable); notOk {
+		return starlark.None, fmt.Errorf("expected argument not to be a function, but was %T", actual)
+	}
+
+	var expectedYamlString string
+	var actualYamlString string
+
+	if expected.Type() == "yamlfragment" {
+		expectedStarlarkValue, err := core.NewStarlarkValue(expected).AsGoValue()
+		if err != nil {
+			return starlark.None, err
+		}
+		document := yamlmeta.Document{Value: expectedStarlarkValue}
+		expectedYaml, err := document.AsYAMLBytes()
+		if err != nil {
+			return starlark.None, err
+		}
+		expectedYamlString = string(expectedYaml)
+	}
+
+	if actual.Type() == "yamlfragment" {
+		actualStarlarkValue, err := core.NewStarlarkValue(actual).AsGoValue()
+		if err != nil {
+			return starlark.None, err
+		}
+		document := yamlmeta.Document{Value: actualStarlarkValue}
+		actualYaml, err := document.AsYAMLBytes()
+		if err != nil {
+			return starlark.None, err
+		}
+		actualYamlString = string(actualYaml)
+	}
+
+	if expected.Type() == "yamlfragment" && actual.Type() != "yamlfragment" {
+		return starlark.None, fmt.Errorf("Not equal:\n"+
+			"expected:\n%s\nactual:\n%s", expectedYamlString, actual.String())
+	} else if expected.Type() != "yamlfragment" && actual.Type() == "yamlfragment" {
+		return starlark.None, fmt.Errorf("Not equal:\n"+
+			"expected:\n%s\n\nactual:\n%s", expected.String(), actualYamlString)
```

Comment:
Nice work handing all of these cases. 

I think it would simplify the number of if statements here if these error cases were instead done in their respective if statements above. For example, in the if statement for: 
[`if expected.Type() == "yamlfragment" {`](https://github.com/vmware-tanzu/carvel-ytt/pull/555/files#diff-d07fbb786694723298677950720360973e65ec6c38222b5f708e839ef53281bbR48)

A check can be added for ` if actual.Type() != "yamlfragment"`, and then we could report this error, instead of doing it here. I Think that would make it easier to follow all of the possible outcomes of the function.